common/remote/rpc/rpc_client.go

```diff
@@ -0,0 +1,508 @@
+	clientAbilities             rpc_request.ClientAbilities
+	Tenant                      string
+}
+
+type ServerRequestHandlerMapping struct {
+	serverRequest func() rpc_request.IRequest
+	handler       IServerRequestHandler
+}
+
+type ReconnectContext struct {
+	onRequestFail bool
+	serverInfo    ServerInfo
+}
+
+type ConnectionEvent struct {
+	eventType ConnectionStatus
+}
+
+func (r *RpcClient) putAllLabels(labels map[string]string) {
+	for k, v := range labels {
+		r.labels[k] = v
+	}
+}
+
+func (r *RpcClient) GetRpcClient() *RpcClient {
+	return r
+}
+
+/**
+ * get all client.
+ *
+ */
+func getAllClient() map[string]IRpcClient {
+	return clientMap
+}
+
+func getClient(clientName string) IRpcClient {
+	return clientMap[clientName]
+}
+
+func CreateClient(clientName string, connectionType ConnectionType, labels map[string]string, nacosServer *nacos_server.NacosServer) (IRpcClient, error) {
+	cMux.Lock()
+	defer cMux.Unlock()
+	if _, ok := clientMap[clientName]; !ok {
+		var rpcClient IRpcClient
+		if GRPC == connectionType {
+			rpcClient = NewGrpcClient(clientName, nacosServer)
+		}
+		if rpcClient == nil {
+			return nil, errors.New("unsupported connection type")
+		}
+		rpcClient.putAllLabels(labels)
+		clientMap[clientName] = rpcClient
+		return rpcClient, nil
+	}
+	return clientMap[clientName], nil
+}
+
+func (r *RpcClient) Start() {
+	if ok := atomic.CompareAndSwapInt32((*int32)(&r.rpcClientStatus), (int32)(INITIALIZED), (int32)(STARTING)); !ok {
+		return
+	}
+	r.registerServerRequestHandlers()
+	r.registerClientResponses()
+	go func() {
+		for {
+			event := <-r.eventChan
+			r.notifyConnectionEvent(event)
+		}
+	}()
+
+	go func() {
+		timer := time.NewTimer(5 * time.Second)
+		for {
+			select {
+			case rc := <-r.reconnectionChan:
+				if (rc.serverInfo != ServerInfo{}) {
+					var serverExist bool
+					for _, v := range r.nacosServer.GetServerList() {
+						if rc.serverInfo.serverIp == v.IpAddr {
+							rc.serverInfo.serverPort = v.Port
+							serverExist = true
+						}
+					}
+					if !serverExist {
+						logger.Infof("%s Recommend server is not in server list ,ignore recommend server %+v", r.name, rc.serverInfo)
+						rc.serverInfo = ServerInfo{}
+					}
+				}
+				r.reconnect(rc.serverInfo, rc.onRequestFail)
+			case <-timer.C:
+				r.healthCheck(timer)
+			}
+		}
+	}()
+
+	var currentConnection IConnection
+	startUpRetryTimes := constant.REQUEST_DOMAIN_RETRY_TIME
+	for startUpRetryTimes > 0 && currentConnection == nil {
+		startUpRetryTimes--
+		serverInfo := r.nextRpcServer()
+
+		logger.Infof("[RpcClient.Start] %+v Try to connect to server on start up, server: %+v", r.name, serverInfo)
+
+		if connection, err := r.executeClient.connectToServer(serverInfo); err != nil {
+			logger.Warnf("[RpcClient.Start] %+v Fail to connect to server on start up, error message=%+v, "+
+				"start up retry times left=%+v", r.name, err.Error(), startUpRetryTimes)
+		} else {
+			currentConnection = connection
```

Comment:
Need to add break here?

Reply:
No, because the condition for the for loop is currentConnection == nil